Add CI

zsusswein opened the pull request:
Adds an initial Github Actions setup. Mainly based on
`usethis::use_tidy_github_actions()` with some changes:

- The R CMD CHECK runs are pared back to only Ubuntu/OSX/Windows latest
  versions
- ~~The codecov check is dropped because it had some setup with a token
  required. Not totally sure what the deal is there, so I replaced it
  with a simple unit test run for now.~~
    - EDIT: Got it working. Left the unit tests as a post-push hook.
- I added a check that NEWS.md is modified with each PR into main, so we
  remember to update it with the changes
- I added in a check that fixup commits are blocked, so we remember to
  autosquash before merging
- Set up the pkgdown deployment

I also added in a placeholder R script and test file to run on the
actions. These will be replaced with real functionality in future PRs.

zsusswein commented:
I've set all but R CMD check to be required. I figured we might see an occasional weird Windows thing or a NOTE on the build that's not worth immediately addressing. We could require everything or require nothing; I don't feel especially strongly.

seabbs commented on .github/workflows/r-unit-tests.yaml:
```diff
@@ -0,0 +1,28 @@
+      - name: Install dependencies
+        uses: r-lib/actions/setup-r-dependencies@v2
+        with:
+          cache: true
```
Is there a reason to set this manually here? If so is there a reason it isn't set everywhere else?

zsusswein replied:
They're the defaults. Removed them to match the other workflow files.

seabbs commented on .github/workflows/r-unit-tests.yaml:
```diff
@@ -0,0 +1,28 @@
+name: Run R Unit Tests
```
Its unusual to have an action for this as they are already being checked in the Rmd check. More usually people would just do this with code coverage

seabbs commented on .github/workflows/pr-commands.yaml:
```diff
@@ -0,0 +1,79 @@
+      - uses: r-lib/actions/setup-r-dependencies@v2
+        with:
+          extra-packages: any::roxygen2
+          needs: pr-document
```
These are special dependency groups you can add to your description if you need them. Looking at your description you don't have any so I would either remove or add. An example for the pkgdown would be the pkgdown package and for linting it would be lintr

seabbs reviewed (commented):
This overall looks good to me. 

I think I would definitely make Rmd checks as required (usually these are the only thing that are required and also sometimes the only CI people have). This is because they are testing the integrity of your package. NOTEs etc shouldn't cause a build failure. 

I also see  some issues with DESCRIPTION groups that don't exist. Finally you likely want to add some badges for these to your README.

> The codecov check is dropped because it had some setup with a token required. Not totally sure what the deal is there, so I replaced it with a simple unit test run for now.

This is an odd one as it should work as long as someone has authorised cdcgov to use codecov. They must have done so as we are using it in the `EpiAware subdirectory`.

seabbs commented:
Whilst we are here we might also want to enable a merge queue so that we have to worry less about rebasing code al the time (it gets rebased and then all CI run in the merge queue and then merged if it passes and bounced if it doesn't)

zsusswein commented:
> I think I would definitely make Rmd checks as required 

Ha, ok! Switched it around so it's the R CMD checks that are required.

>  also see some issues with DESCRIPTION groups that don't exist

Fixed.

> Finally you likely want to add some badges for these to your README.

I think this is a nice-to-have, but would like to leave out of scope for this PR. Moved to #12.

> I agree we can move on but lets make an issue with some detail and aim to get it resolved asap.

Took a little messing around (sorry for all the notifications!) but it's working now.

> we might also want to enable a merge queue

I don't mind manually rebasing right now. Let's revisit if this picks up steam?

seabbs commented on .github/workflows/pre-commit.yaml:
```diff
@@ -1,14 +1,51 @@
-    - uses: actions/checkout@v3
-    - uses: actions/setup-python@v3
-    - uses: ./.github/actions/pre-commit
+      - name: Cancel Previous Runs
```
can you not do this with concurrency settings now and avoid the third-party action?

zsusswein replied:
I imagine I could -- this is the default that `precommit::use_precommit(ci = "gha")` gives you. If you have a nice alternative, I'm happy to switch over.

seabbs replied:
https://github.com/epinowcast/epinowcast/blob/90f7a5a05892a9d176cc71422b796b1c54aa24d6/.github/workflows/render-readme.yaml#L3

seabbs commented on .github/workflows/r-unit-tests.yaml:
```diff
@@ -0,0 +1,25 @@
+
+      - name: Run unit tests
+        run: |
+          testthat::test_dir("tests/testthat/")
```
Can you explain why you want this vs just doing all the testing in r check and code coverage as people normally do?

zsusswein replied:
I like having the tests run post-push on all branches. R CMD check/codecov are set to PR-only.

zsusswein replied:
maybe we should make them run on PR draft rather than post-push?

seabbs replied:
So they will run on PR draft. I think its fine if you want them for all pushes but just make it bee Rmd check and  control wastage with concurrency groups? 

Personally i just make draft Prs to get this and I like that as it encourages me to do so early on if I want CI

seabbs reviewed (commented):
All looking pretty good. A few last questions/comments

zsusswein commented:
> As per CFA policy I think you also want pre-commit to be required

✅ 

>  probably your linting check

Linting is in pre-commit so that should be covered.

seabbs commented on .github/workflows/R-CMD-check.yaml:
```diff
@@ -0,0 +1,46 @@
+# Workflow derived from https://github.com/r-lib/actions/tree/v2/examples
+# Need help debugging build failures? Start at https://github.com/r-lib/actions#where-to-find-help
+on:
```
This should be on push to main as well as should everything else

seabbs commented on .github/workflows/pre-commit.yaml:
```diff
@@ -1,14 +1,51 @@
-  pull_request:
   push:
-    branches: [main]
+    branches-ignore:
```
sorry why would you not run pre-commit on main?

zsusswein replied:
This is another `precommit::use_precommit(ci = "gha")` default

seabbs replied:
I think this one has a case for not being run vs some of the others (as nothing should get to main without a PR. that being said it happens)

zsusswein replied:
I would think even in a hotfix we'd PR and force override branch protection? I'm not sure it's worth gaming this situation out too much because it's pretty extreme.

zsusswein replied:
Fixed.

seabbs reviewed (approved):
LGTM. Oh my is this collapsing of commits not via a squash merge CFA policy? I find it very hard to review final PR changes and I don't really get the reason for it vs a squash merge either.

zsusswein commented:
No, I was just rebasing. I can stop if you prefer a squash at the end.
